lib/ipam/ipam_block_reader_writer.go

```diff
@@ -187,110 +208,146 @@ func (rw blockReaderWriter) claimBlockAffinity(ctx context.Context, subnet cnet.
-	for i := 0; i < ipamEtcdRetries; i++ {
-		// Read the model.KVPair containing the block
-		// and pull out the allocationBlock object.  We need to hold on to this
-		// so that we can pass it back to the datastore on Update.
-		obj, err := rw.client.Get(ctx, model.BlockKey{CIDR: blockCIDR}, "")
-		if err != nil {
-			log.Errorf("Error getting block %s: %v", blockCIDR.String(), err)
-			return err
-		}
-		b := allocationBlock{obj.Value.(*model.AllocationBlock)}
+	// Make sure hostname is not empty.
+	if host == "" {
+		log.Errorf("Hostname can't be empty")
+		return errors.New("Hostname must be sepcified to release block affinity")
+	}
 
-		// Make sure hostname is not empty.
-		if host == "" {
-			log.Errorf("Hostname can't be empty")
-			return errors.New("Hostname must be sepcified to release block affinity")
-		}
+	// Read the model.KVPair containing the block affinity.
+	logCtx := log.WithFields(log.Fields{"block": blockCIDR.String()})
+	logCtx.Debugf("Attempt to release affinity for block")
+	aff, err := rw.client.Get(ctx, model.BlockAffinityKey{Host: host, CIDR: blockCIDR}, "")
+	if err != nil {
+		logCtx.WithError(err).Errorf("Error getting block affinity %s", blockCIDR.String())
+		return err
+	}
 
-		// Check that the block affinity matches the given affinity.
-		if b.Affinity != nil && !hostAffinityMatches(host, b.AllocationBlock) {
-			log.Errorf("Mismatched affinity: %s != %s", *b.Affinity, "host:"+host)
-			return affinityClaimedError{Block: b}
-		}
+	// Mark it as pending deletion.
+	aff.Value.(*model.BlockAffinity).State = model.StatePendingDeletion
+	aff, err = rw.client.Update(ctx, aff)
+	if err != nil {
+		logCtx.WithError(err).Warnf("Failed to mark block affinity as pending deletion")
+		return err
+	}
 
-		if b.empty() {
-			// If the block is empty, we can delete it.
-			_, err := rw.client.Delete(ctx, model.BlockKey{CIDR: b.CIDR}, "")
-			if err != nil {
-				// Return the error unless the block didn't exist.
-				if _, ok := err.(cerrors.ErrorResourceDoesNotExist); !ok {
-					log.Errorf("Error deleting block: %v", err)
-					return err
-				}
-			}
-		} else {
-			// Otherwise, we need to remove affinity from it.
-			// This prevents the host from automatically assigning
-			// from this block unless we're allowed to overflow into
-			// non-affine blocks.
-			b.Affinity = nil
-
-			// Pass back the original KVPair with the new
-			// block information so we can do a CAS.
-			obj.Value = b.AllocationBlock
-			_, err = rw.client.Update(ctx, obj)
-			if err != nil {
-				if _, ok := err.(cerrors.ErrorResourceUpdateConflict); ok {
-					// CASError - continue.
-					continue
-				} else {
-					return err
-				}
-			}
-		}
+	// Read the model.KVPair containing the block
+	// and pull out the allocationBlock object.  We need to hold on to this
+	// so that we can pass it back to the datastore on Update.
+	obj, err := rw.client.Get(ctx, model.BlockKey{CIDR: blockCIDR}, "")
```

Comment:
I _think_ you need to read the block before you do the update to the affinity.  Otherwise another process could slip in, reclaim the block and do an assignment _before_ you read the block here.  Then you'd go ahead and remove the affinity from the block but the CaD for the affinity would fail so we'd have a dangling affinity.